### PR TITLE
Notification/lockscreen area for ios, playback speed, bug fix

### PR DESCRIPTION
**Pull request by subhash279**
Hi,
I was working on some new features for my app and I thought I will merge the changes I made. There are few major features added and one bug fix.

1) bug fix to https://github.com/luanpotter/audioplayers/issues/267

2) Notification area/lock screen on ios. It works both ways - so the notification area can be controlled through the plugin/app and also any changes in the notification area like play/pause, forward/backward(by custom time intervals) will also be reflected in the app. The setNotification function can be called in onDurationChange(I added in example) so that after the audio has loaded, the notification will be set.
ios_notification

3) Change playback speed. This function can be called after play() or resume() or anytime the user in your app wants to change the playback speed.

These features are only on ios for now. I am working on android and will be up in couple of weeks. Please let me know any issue and merge this soon as this might be useful to lot of others.

### CODE REVIEW

**Comment by subhash279** [comment]
I have already been using it in my app after testing. So please let me know if you need anything else.

**Comment by zeucxb** [comment]
It's awesome! How is the android version going?

**Comment by subhash279** [comment]
I am waiting for this PR to be merged before I can work on the android version!

**Comment by ruelluna** [comment]
This will be a great addition!

**Comment by bardiarastin** [comment]
can't wait to see this is landed <3

**Comment by lnalex** [comment]
Cannot wait to be merge ! Well played :1st_place_medal: 

However it's better if you could exclude project files and lock files from committing (project.pbxproj and Podfile.lock). If every contributors change these files, the history will be a mess :) 

Thanks for your contribution!

**Comment by subhash279** [comment]
@lnalex  Thanks for that. I will keep that in mind :)

@erickzanardo Can you pls merge this soon. I don't mean to rush you and I understand this is just a side project. If you can merge this, I can fix a few more bugs and add some other high priority features. This plugin is very important for my project. Keep up the good work!!

File: example/ios/Podfile.lock [hunk 1]
```diff
@@ -6,17 +6,17 @@ PODS:
     - Flutter
 
 DEPENDENCIES:
-  - audioplayers (from `/Users/erick/projects/audioplayers/ios`)
-  - Flutter (from `/Users/erick/flutter/bin/cache/artifacts/engine/ios-release`)
-  - path_provider (from `/Users/erick/flutter/.pub-cache/hosted/pub.dartlang.org/path_provider-1.1.0/ios`)
+  - audioplayers (from `/Users/subhash/Documents/Dev/Podcast/audioplayers/ios`)
```
Comment (luanpotter) [inline example/ios/Podfile.lock]: should this file be committed if it has this absolute path information?

File: example/lib/player_widget.dart [hunk 2]
```diff
@@ -127,10 +127,25 @@ class _PlayerWidgetState extends State<PlayerWidget> {
...
-    _durationSubscription =
-        _audioPlayer.onDurationChanged.listen((duration) => setState(() {
-              _duration = duration;
-            }));
+    _durationSubscription = _audioPlayer.onDurationChanged.listen((duration) {
+      setState(() {
+        _duration = duration;
+      });
+
+      // implemented for ios, android will be done in next few weeks
```
Comment (luanpotter) [inline example/lib/player_widget.dart]: `// TODO implemented for iOS, waiting for android impl`

File: example/lib/player_widget.dart [hunk 3]
```diff
@@ -172,6 +187,15 @@ class _PlayerWidgetState extends State<PlayerWidget> {
     final result =
         await _audioPlayer.play(url, isLocal: isLocal, position: playPosition);
     if (result == 1) setState(() => _playerState = PlayerState.playing);
+
+    // implemented for ios, android will be done in next few weeks
```
Comment (luanpotter) [inline example/lib/player_widget.dart]: same here

File: ios/Classes/AudioplayersPlugin.m [hunk 4]
```diff
@@ -28,6 +29,18 @@ @implementation AudioplayersPlugin {
...
+NSString *_currentPlayerId; // to be used for notifications command center
+MPNowPlayingInfoCenter *_infoCenter;
+MPRemoteCommandCenter *remoteCommandCenter;
+
+NSString *_title; 
+NSString *_albumTitle;
+NSString *_artist;
+NSString *_imageUrl;
+int _duration;
+// int _elapsedTime;
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: no commented out code!

File: ios/Classes/AudioplayersPlugin.m [hunk 5]
```diff
@@ -168,18 +206,160 @@ -(void) initPlayerInfo: (NSString *) playerId {
...
+        duration:  (int) duration
+        elapsedTime:  (int) elapsedTime
+        playerId: (NSString*) playerId {
+  
+  _title = title;
+  _albumTitle = albumTitle;
+  _artist = artist;
+  _imageUrl = imageUrl;
+  _duration = duration;
+  // _elapsedTime = elapsedTime;
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: commented out code!

File: ios/Classes/AudioplayersPlugin.m [hunk 6]
```diff
@@ -168,18 +206,160 @@ -(void) initPlayerInfo: (NSString *) playerId {
...
+
+  _infoCenter = [MPNowPlayingInfoCenter defaultCenter];
+  
+  [ self updateNotification:elapsedTime ];
+
+  if(remoteCommandCenter == nil){
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: `if (...) {` (spacing)

File: ios/Classes/AudioplayersPlugin.m [hunk 7]
```diff
@@ -168,18 +206,160 @@ -(void) initPlayerInfo: (NSString *) playerId {
...
+
+    MPSkipIntervalCommand *skipForwardIntervalCommand = [remoteCommandCenter skipForwardCommand];
+    skipForwardIntervalCommand.preferredIntervals = @[@(forwardSkipInterval)];  // Max 99
+    [skipForwardIntervalCommand setEnabled:YES];
+    [skipForwardIntervalCommand addTarget:self action:@selector(skipForwardEvent:)];
+
+    MPRemoteCommand *pauseCommand = [remoteCommandCenter pauseCommand];
+    [pauseCommand setEnabled:YES];
+    [pauseCommand addTarget:self action:@selector(playOrPauseEvent:)];
+    //
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: just an empty line is enough

File: ios/Classes/AudioplayersPlugin.m [hunk 8]
```diff
@@ -168,18 +206,160 @@ -(void) initPlayerInfo: (NSString *) playerId {
...
+    //    
+    MPRemoteCommand *playCommand = [remoteCommandCenter playCommand];
+    [playCommand setEnabled:YES];
+    [playCommand addTarget:self action:@selector(playOrPauseEvent:)];
+
+    MPRemoteCommand *togglePlayPauseCommand = [remoteCommandCenter togglePlayPauseCommand];
+    [togglePlayPauseCommand setEnabled:YES];
+    [togglePlayPauseCommand addTarget:self action:@selector(playOrPauseEvent:)];
+  }
+
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: extra empty line at end of method

File: ios/Classes/AudioplayersPlugin.m [hunk 9]
```diff
@@ -168,18 +206,160 @@ -(void) initPlayerInfo: (NSString *) playerId {
...
+}
+
+-(void)skipBackwardEvent: (MPSkipIntervalCommandEvent *)skipEvent
+{
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: move the `{` to the preceding line

File: ios/Classes/AudioplayersPlugin.m [hunk 10]
```diff
@@ -168,18 +206,160 @@ -(void) initPlayerInfo: (NSString *) playerId {
...
+    NSLog(@"Skip backward by %f", skipEvent.interval);
+    NSMutableDictionary * playerInfo = players[_currentPlayerId];
+    AVPlayer *player = playerInfo[@"player"];
+    AVPlayerItem *currentItem = player.currentItem;
+    CMTime currentTime = currentItem.currentTime;
+    CMTime newTime = CMTimeSubtract(currentTime, CMTimeMakeWithSeconds(skipEvent.interval, NSEC_PER_SEC));
+    // if CMTime is negative, set it to zero
+    if(CMTimeGetSeconds(newTime) < 0){
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: `if (...) {` (spacing)

File: ios/Classes/AudioplayersPlugin.m [hunk 11]
```diff
@@ -168,18 +206,160 @@ -(void) initPlayerInfo: (NSString *) playerId {
...
+      [ self seek:_currentPlayerId time:CMTimeMakeWithSeconds(0,1) ];
+    } else {
+      [ self seek:_currentPlayerId time:newTime ];
+    }
+}
+
+-(void)skipForwardEvent: (MPSkipIntervalCommandEvent *)skipEvent
+{
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: move the `{` to the preceding line

File: ios/Classes/AudioplayersPlugin.m [hunk 12]
```diff
@@ -168,18 +206,160 @@ -(void) initPlayerInfo: (NSString *) playerId {
...
+    CMTime newTime = CMTimeAdd(currentTime, CMTimeMakeWithSeconds(skipEvent.interval, NSEC_PER_SEC));
+    // if CMTime is more than max duration, limit it
+    if(CMTimeGetSeconds(newTime) > CMTimeGetSeconds(maxDuration)) {
+      [ self seek:_currentPlayerId time:maxDuration ];
+    } else {
+      [ self seek:_currentPlayerId time:newTime ];
+    }
+}
+-(void)playOrPauseEvent: (MPSkipIntervalCommandEvent *)playOrPauseEvent
+{
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: same

File: ios/Classes/AudioplayersPlugin.m [hunk 13]
```diff
@@ -168,18 +206,160 @@ -(void) initPlayerInfo: (NSString *) playerId {
...
+    NSLog(@"playOrPauseEvent");
+
+    NSMutableDictionary * playerInfo = players[_currentPlayerId];
+    AVPlayer *player = playerInfo[@"player"];
+    if (player.timeControlStatus == AVPlayerTimeControlStatusPlaying) {
+        //player is playing and pause it
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: `// ...` (space after `//`)

File: ios/Classes/AudioplayersPlugin.m [hunk 14]
```diff
@@ -168,18 +206,160 @@ -(void) initPlayerInfo: (NSString *) playerId {
...
+        [ self pause:_currentPlayerId ];
+    }
+    else if (player.timeControlStatus == AVPlayerTimeControlStatusPaused) {
+        //player is paused and resume it
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: `// ...` (space after `//`)

File: ios/Classes/AudioplayersPlugin.m [hunk 15]
```diff
@@ -168,18 +206,160 @@ -(void) initPlayerInfo: (NSString *) playerId {
...
+{
+    NSLog(@"playOrPauseEvent");
+
+    NSMutableDictionary * playerInfo = players[_currentPlayerId];
+    AVPlayer *player = playerInfo[@"player"];
+    if (player.timeControlStatus == AVPlayerTimeControlStatusPlaying) {
+        //player is playing and pause it
+        [ self pause:_currentPlayerId ];
+    }
+    else if (player.timeControlStatus == AVPlayerTimeControlStatusPaused) {
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: move that to the preceding line

File: ios/Classes/AudioplayersPlugin.m [hunk 16]
```diff
@@ -168,18 +206,160 @@ -(void) initPlayerInfo: (NSString *) playerId {
...
+        //player is paused and resume it
+        [ self resume:_currentPlayerId ];
+    }
+}
+
+-(void) updateNotification: (int) elapsedTime {
+
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: extra empty line at beginning of method

File: ios/Classes/AudioplayersPlugin.m [hunk 17]
```diff
@@ -168,18 +206,160 @@ -(void) initPlayerInfo: (NSString *) playerId {
...
+  
+  NSMutableDictionary *playingInfo = [NSMutableDictionary dictionary];
+  playingInfo[MPMediaItemPropertyTitle] = _title;
+  playingInfo[MPMediaItemPropertyAlbumTitle] = _albumTitle;
+  playingInfo[MPMediaItemPropertyArtist] = _artist;
+  
+  NSURL *url = [[NSURL alloc] initWithString:_imageUrl];
+  UIImage *artworkImage = [UIImage imageWithData:[NSData dataWithContentsOfURL:url]];
+  if(artworkImage)
+  {
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: `if (...) {`

File: ios/Classes/AudioplayersPlugin.m [hunk 18]
```diff
@@ -168,18 +206,160 @@ -(void) initPlayerInfo: (NSString *) playerId {
...
+      playingInfo[MPMediaItemPropertyArtwork] = albumArt;
+  }
+
+  playingInfo[MPMediaItemPropertyPlaybackDuration] = [NSNumber numberWithInt: _duration];
+  playingInfo[MPNowPlayingInfoPropertyElapsedPlaybackTime] = [NSNumber numberWithInt: elapsedTime];
+
+  playingInfo[MPNowPlayingInfoPropertyPlaybackRate] = @(_playbackRate);
+  NSLog(@"setNotification done");
+
+  if(_infoCenter != nil) {
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: `if[space](`

File: ios/Classes/AudioplayersPlugin.m [hunk 19]
```diff
@@ -168,18 +206,160 @@ -(void) initPlayerInfo: (NSString *) playerId {
...
   NSLog(@"setUrl %@", url);
 
+  // code moved from play() to setUrl() to fix the bug of audio not playing in ios background
+  NSError *error = nil;
+  AVAudioSessionCategory category;
+  if (respectSilence) {
+      category = AVAudioSessionCategoryAmbient;
+  } else {
+      category = AVAudioSessionCategoryPlayback;
+  }
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: I think a ternary is more succinct here ;)
`category = respectSilence ? AVAudioSessionCategoryAmbient : AVAudioSessionCategoryPlayback;`

File: ios/Classes/AudioplayersPlugin.m [hunk 20]
```diff
@@ -245,23 +425,26 @@ -(void) play: (NSString*) playerId
...
-        category = AVAudioSessionCategoryPlayback;
-    }
-    BOOL success = [[AVAudioSession sharedInstance]
-                    setCategory: category
-                    error:&error];
-  if (!success) {
-    NSLog(@"Error setting speaker: %@", error);
-  }
-  [[AVAudioSession sharedInstance] setActive:YES error:&error];
+  // ***** this code is moved to setUrl() to fix the bug of audio not playing in ios background
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: no commented out code! if it was moved somewhere else, remove from here!

File: ios/Classes/AudioplayersPlugin.m [hunk 21]
```diff
@@ -304,12 +487,11 @@ -(void) onTimeInterval: (NSString *) playerId
     if (_isDealloc) {
         return;
     }
-    int mseconds =  CMTimeGetSeconds(time)*1000;
+    int seconds = CMTimeGetSeconds(time);
+    int mseconds =  seconds*1000;
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: extra space here (two spaces in a row)

File: ios/Classes/AudioplayersPlugin.m [hunk 22]
```diff
@@ -335,6 +518,25 @@ -(void) setVolume: (float) volume
   [ player setVolume:volume ];
 }
 
+-(void) setPlaybackRate: (float) playbackRate 
+        playerId:  (NSString *) playerId {
+  NSLog(@"ios -> calling setPlaybackRate");
+  
+  NSMutableDictionary *playerInfo = players[playerId];
+  AVPlayer *player = playerInfo[@"player"];
+  // playerInfo[@"volume"] = @(volume);
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: commented out code

File: ios/Classes/AudioplayersPlugin.m [hunk 23]
```diff
@@ -335,6 +518,25 @@ -(void) setVolume: (float) volume
...
+  _playbackRate = playbackRate;
+  [ player setRate:playbackRate ];
+
+  if(_infoCenter != nil) {
+    // NSMutableDictionary * playerInfo = players[playerId];
+    // AVPlayer *player = playerInfo[@"player"];
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: commented out code

File: ios/Classes/AudioplayersPlugin.m [hunk 24]
```diff
@@ -335,6 +518,25 @@ -(void) setVolume: (float) volume
...
+        playerId:  (NSString *) playerId {
+  NSLog(@"ios -> calling setPlaybackRate");
+  
+  NSMutableDictionary *playerInfo = players[playerId];
+  AVPlayer *player = playerInfo[@"player"];
+  // playerInfo[@"volume"] = @(volume);
+  _playbackRate = playbackRate;
+  [ player setRate:playbackRate ];
+
+  if(_infoCenter != nil) {
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: `if (`

File: ios/Classes/AudioplayersPlugin.m [hunk 25]
```diff
@@ -356,6 +558,11 @@ -(void) seek: (NSString *) playerId
   NSMutableDictionary * playerInfo = players[playerId];
   AVPlayer *player = playerInfo[@"player"];
   [[player currentItem] seekToTime:time];
+
+  int seconds = CMTimeGetSeconds(time);
+  if(_infoCenter != nil) {
```
Comment (luanpotter) [inline ios/Classes/AudioplayersPlugin.m]: `if (`

File: lib/audioplayers.dart [hunk 26]
```diff
@@ -324,14 +324,46 @@ class AudioPlayer {
...
 
+  /// Sets the playback rate - call this after first calling play() or resume()
+  ///
+  /// ios has limits between 0.5 and 2x
+  /// not sure if that's changed recently.
+  Future<int> setPlaybackRate({double playbackRate = 1.0}) {
+    return _invokeMethod('setPlaybackRate', {'playbackRate': playbackRate});
+  }
+
+  /// Sets the notification bar for lock screen and notification area in ios for now.
```
Comment (luanpotter) [inline lib/audioplayers.dart]: iOS

File: lib/audioplayers.dart [hunk 27]
```diff
@@ -324,14 +324,46 @@ class AudioPlayer {
     );
   }
 
+  /// Sets the playback rate - call this after first calling play() or resume()
```
Comment (luanpotter) [inline lib/audioplayers.dart]: add a warning that this only works for iOS for now

**Comment by subhash279** [comment]
@luanpotter thanks for the comments. I will fix these and submit asap.
I understand this is a general purpose audio plugin and that's why the new code was written mostly independently of old code in such a way that the notification area shows only when the function setNotification() is expilictly called on onDurationChanged() otherwise it doesn't show up and the plugin works as usual. I think many people here would be using this plugin for media players and this feature would be really useful for them.

**Comment by subhash279** [comment]
@luanpotter Made all the changes. Please let me know if there is anything else needed.

**Comment by bardiarastin** [comment]
@subhash279 hi, thanks for awesome work, I wanted to know that when will you start working on android too? one more thing is that this lib does not take array of songs so how we can achieve handling next and previous buttons on notification bar? 
maybe you could add a onRequestNext call back so we can handle changing song when the user presses the next button on notification bar.

**Comment by subhash279** [comment]
@bardiarastin this is just the beginning of the implementation for notification area. On iOS, this can be used to play/pause and forward/backward any audio by specified time interval. I will be working on skipping to the next/previous audio in the playlist soon. I wanted to make sure this gets merged without any issues so that I could start working on more features and also for android. I will merge when that is done but it will take a few weeks atleast.

**Comment by erickzanardo** [comment]
@subhash279 don't you think you can submit the Android before we merge, a nice feature like that is not cool to be added without both platforms. Either, maybe we can merge this without having the Android part, but for releasing a new version, with this, we probably would wait for the android implementation. What do you think @luanpotter ?

**Comment by luanpotter** [comment]
I think if @subhash279 is already working on the android version, we can wait. Maybe upon android implementation you discover you better tweak the API a bit because of some unforeseen event, and then you will have full control over that before merging.

File: example/lib/player_widget.dart [hunk 28]
```diff
@@ -127,10 +127,25 @@ class _PlayerWidgetState extends State<PlayerWidget> {
...
     _audioPlayer = AudioPlayer(mode: mode);
 
-    _durationSubscription =
-        _audioPlayer.onDurationChanged.listen((duration) => setState(() {
-              _duration = duration;
-            }));
+    _durationSubscription = _audioPlayer.onDurationChanged.listen((duration) {
+      setState(() {
+        _duration = duration;
+      });
```
Comment (luanpotter) [inline example/lib/player_widget.dart]: inline this arrow function: `setState(() => _duration = duration);`

**Comment by subhash279** [comment]
@erickzanardo Can you please merge this first. I think it will be useful for people using it already in iOS apps and also once more devs start using it, we can fine tune any issues and then I can implement them directly in the android version. Thats's my plan.

@luanpotter I made the formatting changes suggested!

**Comment by lnalex** [comment]
We need a little bit of documentation, I think so :)

My app is crashing when clicking on a button (`audioPlayer.setNotification(title: "Title");`)

```
        kMRMediaRemoteNowPlayingInfoDuration = 0
	kMRMediaRemoteNowPlayingInfoAlbum = 
	kMRMediaRemoteNowPlayingInfoElapsedTime = 0
	kMRMediaRemoteNowPlayingInfoContentItemIdentifier = 477926D1-B99B-4D8F-8EE0-1C75B3BA5752
	kMRMediaRemoteNowPlayingInfoArtist = 
	kMRMediaRemoteNowPlayingInfoPlaybackRate = 1
	kMRMediaRemoteNowPlayingInfoTitle = Title
	kMRMediaRemoteNowPlayingInfoTimestamp = 2019-09-24 19:02:43 +0000
}>
 item:<(null)>
 -> 
item:<477926D1-B99B-4D8F-8EE0-1C75B3BA5752/Title/CalculatedPlaybackPosition: 0.002757/PlaybackRate: 1.000000>
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff23b98bde __exceptionPreprocess + 350
	1   libobjc.A.dylib                     0x00007fff503b5b20 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff23b98958 +[NSException raise:format:arguments:] + 88
	3   Foundation                          0x00007fff255eb6f5 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 191
	4   MediaPlayer                         0x00007fff278e90b0 -[MPRemoteCommand addTarget:action:] + 1154
	5   audioplayers                        0x0000000103593b1d -[AudioplayersPlugin setNotification:albumTitle:artist:imageUrl:forwardSkipInterval:backwardSkipInterval:duration:elapsedTime:playerId:] + 621
	6   audioplayers              <…>
```

**Comment by subhash279** [comment]
@lnalex where are you calling audioPlayer.setNotification(title: "Title")?
It should be called only inside onDurationChanged() callback as shown in the repo example. This is to make sure the audio is loaded first before setting it in the notification area.

**Comment by lnalex** [comment]
Even in onDurationChanged() callback it's crashing:

```
_platformCallHandler call audio.onDuration {playerId: c54644e7-4044-4421-bcf0-c599223c17da, value: 1847170}
	kMRMediaRemoteNowPlayingInfoDuration = 1847
	kMRMediaRemoteNowPlayingInfoAlbum = 
	kMRMediaRemoteNowPlayingInfoElapsedTime = 0
	kMRMediaRemoteNowPlayingInfoContentItemIdentifier = 6352B3ED-5E8F-4564-9F8C-92C031252C8C
	kMRMediaRemoteNowPlayingInfoArtist = 
	kMRMediaRemoteNowPlayingInfoPlaybackRate = 1
	kMRMediaRemoteNowPlayingInfoTitle = Title
	kMRMediaRemoteNowPlayingInfoTimestamp = 2019-09-24 19:57:25 +0000
}>
 item:<(null)>
 -> 
item:<6352B3ED-5E8F-4564-9F8C-92C031252C8C/Title/CalculatedPlaybackPosition: 0.002077/PlaybackRate: 1.000000>
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff23b98bde __exceptionPreprocess + 350
	1   libobjc.A.dylib                     0x00007fff503b5b20 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff23b98958 +[NSException raise:format:arguments:] + 88
	3   Foundation                          0x00007fff255eb6f5 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 191
	4   MediaPlayer                         0x00007fff278e90b0 -[MPRemoteCommand addTarget:action:] + 1154
	5   audioplayers                        0x0000000111fa9b1d -[AudioplayersPlugin setNotification:albumTitle:artist:imageUrl:forwardSkipInterval:backwardSkipInterval:duration:elapsedTime:playerId:] + 621
	6   audioplayers              <…>

```

My code:

```
_audioPlayer.onDurationChanged.listen((Duration d) {
      _audioPlayer.setNotification(title: "Title", duration: d);
    });
```

**Comment by subhash279** [comment]
From the error report you pasted, it might be kMRMediaRemoteNowPlayingInfoAlbum and  kMRMediaRemoteNowPlayingInfoArtist are not getting any values. In the setNotification function, can you set albumTitle and artist to non-null and non-empty strings and try again.

**Comment by lnalex** [comment]
I've set all arguments:
```
flutter: _platformCallHandler call audio.onDuration {playerId: 9930437d-2c47-4f9c-9e8a-eeffd04a4aba, value: 1847170}
	kMRMediaRemoteNowPlayingInfoDuration = 1847
	kMRMediaRemoteNowPlayingInfoAlbum = AlbumTitle
	kMRMediaRemoteNowPlayingInfoTimestamp = 2019-09-25 10:20:50 +0000
	kMRMediaRemoteNowPlayingInfoArtworkMIMEType = image/jpeg
	kMRMediaRemoteNowPlayingInfoTitle = Title
	kMRMediaRemoteNowPlayingInfoArtworkIdentifier = 0F08BCE2-045C-4336-AB84-656920C75A99
	kMRMediaRemoteNowPlayingInfoElapsedTime = 0
	kMRMediaRemoteNowPlayingInfoPlaybackRate = 1
	kMRMediaRemoteNowPlayingInfoArtworkDataHeight = 768
	kMRMediaRemoteNowPlayingInfoContentItemIdentifier = 3B6CA692-CEB7-4CE9-B163-300C1E312917
	kMRMediaRemoteNowPlayingInfoArtist = Artist
	kMRMediaRemoteNowPlayingInfoArtworkDataWidth = 768
}>
 item:<(null)>
 -> 
item:<3B6CA692-CEB7-4CE9-B163-300C1E312917/Title/CalculatedPlaybackPosition: 0.002851/PlaybackRate: 1.000000>
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff23b98bde __exceptionPreprocess + 350
	1   libobjc.A.dylib                     0x00007fff503b5b20 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff23b98958 +[NSException raise:format:arguments:] + 88
	3   Foundation                          0x00007fff255eb6f5 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 191
	4   MediaPlayer                         0x00007fff278e90b0 -[MPRemoteCommand addTarget:action:] + 1154
	5   audioplayers                        0x000000010d79fb1d -[AudioplayersPlugin setNotification:albumTitle:artist:imageUrl:forwardSkipInterval:backwardSkipInterval:duration:elapsedTime:playerId:] + 621
	6   audioplayers              <…>
```

**Comment by shadyshrif** [comment]
@subhash279 I would like to thank you for your great effort. I'm itching to get this feature working on android also. Could you please share with us your time expectations for that.

Thanks in advance.

**Comment by subhash279** [comment]
Hi, I am a little busy at the moment but I plan to get it done in about 3-4 weeks!

**Comment by shadyshrif** [comment]
@subhash279 I really appreciate your effort. is there any chance to get the android support for this feature soon.

**Comment by dvorapa** [comment]
@proteye is working on it: https://github.com/luanpotter/audioplayers/pull/304

**Comment by muarachmann** [comment]
@subhash279 this is wonderful, just wondering if the controls could be actually next and previous buttons as it is when playing audios on ios? Any ideas on how to do this?

**Comment by subhash279** [comment]
Hi @muarachmann I am really packed right now but the corresponding code to that would be in AudioplayersPlugin.m file if you want to take a look at it. 

It would be something like this (this code is for forwardTime).
MPSkipIntervalCommand *skipForwardIntervalCommand = [remoteCommandCenter skipForwardCommand];

Similarly you need to add functions for next and previous buttons for them to show up on control screen.

**Pull request merged**